File: C++/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        vector<char> stack;
        for (char c : s) {
            switch (c) {
                case '(': stack.push_back(')'); break;
                case '[': stack.push_back(']'); break;
                case '{': stack.push_back('}'); break;
                case ')':
                case ']':
                case '}':
                    if (!stack.empty() && stack.back() == c)
                    {
                        stack.pop_back();
                    }
                    else
                    {
                        return false;
                    }
                    break;
            }
        }
        return stack.empty();
    }
};

File: Kotlin/valid-parentheses.kt
class Solution {
    fun isValid(s: String): Boolean {
        val stack = ArrayDeque<Char>()
        for (c in s) {
            if (c == '(' || c == '[' || c == '{') {
                stack.add(when (c) { 
                    '(' -> ')'
                    '[' -> ']'
                    '{' -> '}'
                    else -> '*'
                })
            } else if (c == ')' || c == ']' || c == '}') {
                if (!stack.isEmpty() && stack.last() == c) 
                    stack.removeLast()
                else 
                    return false 
            }
        }
        return stack.isEmpty()
    }
}